test/e2e/firewall.go

```diff
@@ -0,0 +1,410 @@
+								fmt.Sprintf("--http-port=%d", httpPort),
+								fmt.Sprintf("--udp-port=%d", udpPort),
+							},
+							Ports: []api.ContainerPort{
+								{
+									Name:          "http",
+									ContainerPort: httpPort,
+								},
+								{
+									Name:          "udp",
+									ContainerPort: udpPort,
+								},
+							},
+						},
+					},
+					SecurityContext: &api.PodSecurityContext{
+						HostNetwork: hostNetwork,
+					},
+				},
+			},
+		},
+	})
+	return err
+}
+
+func deleteDaemonSet(f *framework.Framework, dsName, ns string, labels map[string]string) {
+	framework.Logf("Check that reaper kills all daemon pods for %s", dsName)
+	dsReaper, err := kubectl.ReaperFor(extensions.Kind("DaemonSet"), f.ClientSet)
+	Expect(err).NotTo(HaveOccurred())
+	err = dsReaper.Stop(ns, dsName, 0, nil)
+	Expect(err).NotTo(HaveOccurred())
+	err = wait.Poll(dsRetryPeriod, dsRetryTimeout, checkRunningOnNoNodes(f, labels))
+	Expect(err).NotTo(HaveOccurred(), "error waiting for daemon pod to be reaped")
+}
+
+func getNonMaterNodeNames(c clientset.Interface) (sets.String, error) {
+	masterName := framework.TestContext.CloudConfig.MasterName
+	nodeList, err := c.Core().Nodes().List(api.ListOptions{})
+	if err != nil {
+		return nil, err
+	}
+	nodeNames := sets.NewString()
+	for _, node := range nodeList.Items {
+		if node.Name != masterName {
+			nodeNames.Insert(node.Name)
+		}
+	}
+	return nodeNames, nil
+}
+
+func examineHitNodesFromOutside(externalIP string, httpPort int32, timeout time.Duration, expectedHosts sets.String, exitWhenSatisfy bool) error {
+	hittedHosts := sets.NewString()
+
+	framework.Logf("Waiting up to %v for satisfying expectedHosts", timeout)
+	condition := func() (bool, error) {
+		respBody, err := queryFromUrl(externalIP, "hostname", httpPort, firewallTestHttpTimeout)
+		if err != nil {
+			framework.Logf("%v", err)
+			return false, nil
+		}
+		hittedHost := strings.TrimSpace(respBody)
+		hittedHosts.Insert(hittedHost)
+		if !expectedHosts.Has(hittedHost) {
+			return false, fmt.Errorf("error hitting unexpected host: %v", hittedHost)
+		}
+
+		framework.Logf("Missing %+v, got %+v", expectedHosts.Difference(hittedHosts), hittedHosts)
+		if exitWhenSatisfy && hittedHosts.Equal(expectedHosts) {
+			return true, nil
+		}
+		return false, nil
+	}
+
+	if err := wait.Poll(time.Second, timeout, condition); err != nil {
+		if !exitWhenSatisfy && hittedHosts.Equal(expectedHosts) {
+			return nil
+		}
+		return fmt.Errorf("error waiting for expectedHosts: %v, hittedHosts: %v, err message: %v", expectedHosts, hittedHosts, err)
+	}
+	return nil
+}
+
+func examineNotHitNodeFromOutside(externalIP string, httpPort int32, timeout time.Duration, targetHost string, countToSucceed int) error {
+	count := 0
+
+	framework.Logf("Waiting up to %v for not hitting %v for %v times", timeout, targetHost, countToSucceed)
+	condition := func() (bool, error) {
+		respBody, err := queryFromUrl(externalIP, "hostname", httpPort, firewallTestHttpTimeout)
```

Comment:
I think we're alreayd doing this in various places (hitting externalip:nodePort/url from a jenkins builder). Can you either use the same pattern as those places or try to centralize the logic for reuse? eg: https://github.com/kubernetes/kubernetes/blob/master/test/e2e/service.go#L1242